File: python/day16.py
from aoc import *
import re
from collections import defaultdict
from itertools import product, combinations


pattern = re.compile(r"[A-Z]{2}|\d+")


def parse_line(line):
    valve, flow, *connections = re.findall(pattern, line)
    return valve, int(flow), connections

def create_connections(data):
    valves = set()
    flows = dict()
    connections = defaultdict(lambda: inf)
    for valve, flow, conns in data:
        valves.add(valve)
        if flow or valve == 'AA':
            flows[valve] = flow
        for conn in conns:
            connections[valve, conn] = 1
    for a, b, c in product(valves, repeat=3): # Floyd-Warshall
        if a != b != c != a:
            connections[b, c] = min(connections[b, c],
                                    connections[b, a] + connections[a, c])
    return flows, connections


def score(flows, attempt):
    return sum(flows[valve] * time for valve, time in attempt.items())


def find_best_scores(flows, attempts):
    top_scores = defaultdict(int)
    for attempt in attempts:
        k = frozenset(attempt)
        v = score(flows, attempt)
        top_scores[k] = max(top_scores[k], v)
    return top_scores


def best_tandem_score(flows, attempts):
    top_scores = find_best_scores(flows, attempts)
    return max(h_score + e_score
               for (h_vis, h_score), (e_vis, e_score) in combinations(top_scores.items(), 2)
               if not h_vis & e_vis)


def traverse(time, closed_valves, conns, curr='AA', visited=dict()):
    for valve in closed_valves:
        if (t := time - conns[curr, valve] - 1) > 1:
            yield from traverse(t, closed_valves - {valve}, conns, valve, visited | {valve: t})
    yield visited


def solve(filename=16):
    flows, conns  = create_connections(map(parse_line, read_input(filename)))
    closed_valves = set(flows.keys())
    p1 = max(score(flows, attempt) for attempt in traverse(30, closed_valves, conns))
    p2 = best_tandem_score(flows, traverse(26, closed_valves, conns))
    return p1, p2


print(solve())










def test():
    assert solve("16_test") == (1651, 1707)
    assert solve() == (1880, 2520)

File: clojure/day16.clj
(ns day16
  (:require aoc
            [clojure.data.int-map :refer [int-map]]))


(defn parse-valve [valve]
  (reduce
   (fn [^long acc l]
     (+ (* 26 acc)
        (- (int l) 65)))
   0
   valve))

(defn connection-hash ^long [^long v1 ^long v2]
  (+ (* 26 26 v1) v2))

(defn parse-line [line]
  (let [[valve flow & connections] (re-seq #"[A-Z]{2}|\d+" line)]
    [(parse-valve valve) (parse-long flow) connections]))

(defn create-connections [input]
  (reduce
   (fn [[flows connections] [valve flow conns]]
     [(assoc flows valve flow)
      (reduce #(assoc %1 (connection-hash valve (parse-valve %2)) 1)
              connections
              conns)])
   [(int-map) (int-map)]
   input))

(defn floyd-warshall [valves connections]
  (reduce
   (fn [conns [a b c]]
     (let [bc (connection-hash b c)
           ba (connection-hash b a)
           ac (connection-hash a c)]
       (if-not (distinct? a b c) conns
               (assoc conns bc
                      (min (conns bc 999)
                           (+ (conns ba 999)
                              (conns ac 999)))))))
   connections
   (for [a valves , b valves , c valves] [a b c])))



(defn traverse [current closed-valves conns visited time]
  (into [visited]
        cat
        (for [valve closed-valves
              :let [t (- ^long time
                         ^long (conns (connection-hash current valve))
                         1)]
              :when (> t 1)
              :let [visited' (assoc visited valve t)]]
          (if (< t 6)  ; optimization
            [visited']
            (traverse valve (disj closed-valves valve) conns visited' t)))))


(defn score [flows attempt]
  (transduce
   (map (fn [[valve time]]
          (* ^long (flows valve) ^long time)))
   +
   attempt))

(defn best-score-per-valve-set [flows attempts]
  (->> attempts
       (map (fn [attempt]
              [(set (keys attempt)) (score flows attempt)]))
       (reduce (fn [top-scores [k v]]
                 (assoc! top-scores k (max (top-scores k 0) v)))
               (transient {}))
       persistent!))

(defn tandem-scores [top-scores]
  (for [[^long i [h_vis ^long h_score]] (map-indexed vector top-scores)
        [^long j [e_vis ^long e_score]] (map-indexed vector top-scores)
        :while (> i j)
        :when (aoc/none? h_vis e_vis)]
    (+ h_score e_score)))



(defn part-1 [[valves conns flows]]
  (->> (traverse 0 valves conns {} 30)
       (transduce
        (map #(score flows %))
        max 0)))

(defn part-2 [[valves conns flows]]
  (->> (traverse 0 valves conns {} 26)
       (best-score-per-valve-set flows)
       (sort-by val)
       (take-last 256) ; narrow down the search space
       tandem-scores
       (reduce max)))

(defn parse-input [input]
  (let [[all-flows direct-conns] (create-connections
                                  (aoc/parse-input input parse-line))
        conns  (floyd-warshall (keys all-flows) direct-conns)
        flows (into {} (filter #(pos? (val %))) all-flows)
        valves (set (keys flows))]
    [valves conns flows]))

(defn solve
  ([] (solve (aoc/read-file 16)))
  ([input]
   (let [data (parse-input input)
         p1 (future (part-1 data))
         p2 (future (part-2 data))]
     [@p1 @p2])))


(solve)
